File: api/src/main/java/io/perryquotes/api/quote/author/AuthorController.java
package io.perryquotes.api.quote.author;

import io.perryquotes.api.base.LoggableComponent;
import org.slf4j.Logger;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.util.UriComponentsBuilder;

import javax.validation.Valid;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@RestController
public class AuthorController extends LoggableComponent {

  private final AuthorService authorService;

  public AuthorController(final Logger log,
                          final AuthorService authorService) {
    super(log);
    this.authorService = authorService;
  }

  @GetMapping("/authors/{uuid}")
  public ResponseEntity<AuthorRecord> findByUuid(@PathVariable final UUID uuid) {
    return authorService.findByUuid(uuid)
      .map(author -> ResponseEntity.ok(author.toDTO()))
      .orElse(ResponseEntity.notFound().build());
  }

  @GetMapping("/authors/search")
  public ResponseEntity<AuthorRecord> findByName(@RequestParam final String name) {
    return authorService.findByName(name)
      .map(author -> ResponseEntity.ok(author.toDTO()))
      .orElse(ResponseEntity.notFound().build());
  }

  @GetMapping("/authors")
  public ResponseEntity<List<AuthorRecord>> findAll() {
    var result = authorService.findAll()
      .stream()
      .map(Author::toDTO)
      .collect(Collectors.toList());
    return ResponseEntity.ok(result);
  }

  @PostMapping("/authors")
  public ResponseEntity<AuthorRecord> create(@Valid @RequestBody final AuthorRecord request,
                                                 final UriComponentsBuilder uriBuilder)
  {
    var created = authorService.create(request);
    var location = uriBuilder.path("/authors/{uuid}").buildAndExpand(created.getUuid()).toUri();
    log.info("Created Author at location '{}': {}}", location, created);
    return ResponseEntity.created(location).body(created.toDTO());
  }

  @PutMapping("/authors/{uuid}")
  public ResponseEntity<AuthorRecord> update(@PathVariable final UUID uuid,
                                                 @RequestBody final AuthorRecord request) {
    var updated = authorService.update(uuid, request);
    log.info("Updated Author {}: {}", uuid, updated);
    return ResponseEntity.ok(updated.toDTO());
  }
}

File: api/src/main/kotlin/io/perryquotes/api/author/AuthorController.kt
package io.perryquotes.api.author

import io.perryquotes.core.author.AuthorNotFoundException
import io.perryquotes.core.author.AuthorRequest
import io.perryquotes.core.author.AuthorService
import jakarta.validation.Valid
import mu.KLogging
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.util.UriComponentsBuilder
import java.util.*

@RestController
class AuthorController(private val authorService: AuthorService) {

    companion object : KLogging()

    //PUBLIC
    @GetMapping("/authors/{uuid}")
    fun getByUuid(@PathVariable uuid: UUID): ResponseEntity<AuthorView> {
        logger.debug { "Handle request GET /authors/$uuid" }
        val responseEntity = authorService.getByUuid(uuid)?.let {
            ResponseEntity.ok(it.toView())
        } ?: throw AuthorNotFoundException("Author[uuid=$uuid] not found")
        return responseEntity
    }

    //PUBLIC
    @GetMapping("/authors")
    fun getAll(): ResponseEntity<List<AuthorView>> {
        logger.debug { "Handle request GET /authors" }
        return ResponseEntity.ok(authorService.getAll().map { it.toView() })
    }

    //USER, ADMIN
    @PostMapping("/authors")
    fun createAuthor(
        @Valid @RequestBody request: AuthorRequest,
        uriBuilder: UriComponentsBuilder,
    ): ResponseEntity<AuthorView> {
        logger.debug { "Handle request POST /authors, body=$request" }
        val created = authorService.create(request)
        val location = uriBuilder.path("/authors/{uuid}").buildAndExpand(created.uuid).toUri()
        logger.info { "Created Author=$created at location $location" }
        return ResponseEntity.created(location).body(created.toView())
    }

    //USER, ADMIN
    @PutMapping("/authors/{uuid}")
    fun updateAuthor(
        @PathVariable uuid: UUID,
        @Valid @RequestBody request: AuthorRequest,
    ): ResponseEntity<AuthorView>? {
        logger.debug { "Handle request PUT /authors/$uuid, body=$request" }
        val updated = authorService.update(uuid, request)
        logger.info { "Updated Author[uuid=$uuid]: $updated" }
        return ResponseEntity.ok(updated.toView())
    }

    //ADMIN
    @DeleteMapping("/authors/{uuid}")
    fun deleteAuthor(@PathVariable uuid: UUID): ResponseEntity<Unit> {
        logger.debug { "Handle request DELETE /authors/$uuid" }
        val deleted = authorService.delete(uuid)
        logger.info { "Deleted Author[uuid=$uuid]: $deleted" }
        return ResponseEntity.noContent().build()
    }
}
